### PR TITLE
[diffusion] Initial XPU support for sglang diffusion

### PR DESCRIPTION
**Pull request by xiangyuT**
## Motivation

Add Intel XPU (GPU) support to SGLang Diffusion, enabling image and video generation on Intel GPUs (e.g., Intel Arc Pro B60).

### Tested Models
| Model | Type |
|---|---|
| `Tongyi-MAI/Z-Image-Turbo` | Text-to-Image |
| `black-forest-labs/FLUX.1-dev` | Text-to-Image |
| `black-forest-labs/FLUX.2-klein-4B` | Text-to-Image |
| `black-forest-labs/FLUX.2-klein-9B` | Text-to-Image |
| `Wan-AI/Wan2.1-T2V-1.3B-Diffusers` | Text-to-Video |
| `Wan-AI/Wan2.2-TI2V-5B-Diffusers` | Text/Image-to-Video |

### 1. XPU Platform Abstraction
- **New file:** `platforms/xpu.py` — `XpuPlatform` with device capability detection, memory management, and attention backend selection.
- Updated `platforms/__init__.py` and `platforms/interface.py` to register XPU in the platform auto-detection chain and add the `XPU_FLASH_ATTN` attention backend enum.

### 2. Distributed Communication (XCCL)
- **New file:** `device_communicators/xpu_communicator.py` — `XpuCommunicator` using PyTorch XCCL backend for Intel XPU collective operations.
- Updated `distributed_init_method` to select XCCL backend when running on XPU.

### 3. XPU Flash Attention Backend
- **New file:** `attention/backends/xpu_flash_attn.py` — `XpuFlashAttentionBackend` / `XpuFlashAttentionImpl`.
- Falls back to `torch.nn.functional.scaled_dot_product_attention` when sgl-kernel is unavailable or the device does not support flash attention. Also can be configured by setting `--attention-backend TORCH_SDPA` or `--attention-backend XPU_FLASH_ATTN`.

## Accuracy Tests

> All tested models produce visually correct outputs on Intel Arc Pro B60 GPUs. Sample outputs attached in comments below.

Results
---

- Tongyi-MAI/Z-Image-Turbo

```bash
sglang generate --model-path "/llm/models/Z-Image-Turbo"   --prompt "A curious raccoon peers through a vibrant field of yellow sunflowers, its eyes wide with interest"   --seed 0   --profile --vae-cpu-offload --pin-cpu-memory --text-encoder-cpu-offload --num-gpus 2 --tp-size 2
```
Tongyi-MAI:Z-Image-Turbo

- black-forest-labs/FLUX.1-dev

```bash
sglang generate --model-path "/llm/models/FLUX.1-dev"   --prompt "A curious raccoon peers through a vibrant field of yellow sunflowers, its eyes wide with interest"   --seed 0 --vae-cpu-offload --vae-tiling  --pin-cpu-memory --text-encoder-cpu-offload --num-gpus 4 --tp-size 4 --attention-backend TORCH_SDPA
```

image

- black-forest-labs/FLUX.2-klein 4B

```bash
sglang generate --model-path "/llm/models/FLUX.2-klein-4B"   --prompt "A curious raccoon peers through a vibrant field of yellow sunflowers, its eyes wide with interest"   --seed 0   --profile --vae-cpu-offload --pin-cpu-memory --text-encoder-cpu-offload --num-gpus 2 --tp-size 2
```

black-forest-labs:FLUX 2-klein 4B

- black-forest-labs/FLUX.2-klein 9B

```bash
sglang generate --model-path "/llm/models/FLUX.2-klein-9B"   --prompt "A curious raccoon peers through a vibrant field of yellow sunflowers, its eyes wide with interest"   --seed 0   --profile --vae-cpu-offload --pin-cpu-memory --text-encoder-cpu-offload --num-gpus 2 --tp-size 2
```

black-forest-labs:FLUX 2-klein 9B

- Wan-AI/Wan2.1-T2V-1.3B-Diffusers

```bash
sglang generate --prompt 'A curious raccoon peers through a vibrant field of yellow sunflowers, its eyes wide with interest' --save-output --log-level debug --output-path outputs --model-path "/llm/models/Wan2.1-T2V-1.3B-Diffusers" --num-gpus 2 --tp-size 2 --height 480 --width 832 --pin-cpu-memory --dit-cpu-offload --text-encoder-cpu-offload --vae-cpu-offload --vae-tiling
```

https://github.com/user-attachments/assets/e3eb6246-1265-4991-be8c-89994545b0b8

- Wan-AI/Wan2.2-TI2V-5B-Diffusers

```bash
sglang generate --prompt 'A curious raccoon peers through a vibrant field of yellow sunflowers, its eyes wide with interest' --save-output --log-level debug --output-path outputs --model-path "/llm/models/Wan2.2-TI2V-5B-Diffusers" --num-gpus 2 --tp-size 2 --height 480 --width 832 --pin-cpu-memory --dit-cpu-offload --text-encoder-cpu-offload --vae-cpu-offload --vae-tiling --attention-backend TORCH_SDPA
```

https://github.com/user-attachments/assets/326a7116-dc7f-4fef-bf84-cbcf76020c4e

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/platforms/interface.py [hunk 1]
```diff
@@ -36,6 +36,7 @@ class AttentionBackendEnum(enum.Enum):
     AITER = enum.auto()
     SLA_ATTN = enum.auto()
     SAGE_SLA_ATTN = enum.auto()
+    XPU_FLASH_ATTN = enum.auto()
```
Comment (airMeng) [inline python/sglang/multimodal_gen/runtime/platforms/interface.py]: suggest to use the same naming ```INTEL_XPU```

File: python/sglang/multimodal_gen/runtime/cache/cache_dit_integration.py [hunk 2]
```diff
@@ -87,8 +88,16 @@ def patched_similarity(self, t1, t2, *, threshold, parallelized=False, prefix="F
...
 
-        dist.all_reduce(mean_diff, op=dist.ReduceOp.AVG, group=target_group)
-        dist.all_reduce(mean_t1, op=dist.ReduceOp.AVG, group=target_group)
+        # Use SUM + divide instead of AVG on XPU, because XCCL in torch 2.9+xpu have issues for AVG
+        if current_platform.is_xpu():
+            world_size = dist.get_world_size(target_group)
+            dist.all_reduce(mean_diff, op=dist.ReduceOp.SUM, group=target_group)
+            dist.all_reduce(mean_t1, op=dist.ReduceOp.SUM, group=target_group)
+            mean_diff = mean_diff / world_size
+            mean_t1 = mean_t1 / world_size
```
Comment (mingfeima) [inline python/sglang/multimodal_gen/runtime/cache/cache_dit_integration.py:97]: @zhangxiaoli73 do we have a fix on torch 2.10?

File: python/sglang/multimodal_gen/runtime/distributed/device_communicators/xpu_communicator.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+            # Use ft_c.all_to_all_single instead of dist.all_to_all_single
+            # to avoid XCCL data corruption bug
+            input_shape = input_t.shape
+            output = ft_c.all_to_all_single(
+                input_t.flatten(),
+                output_split_sizes=None,
+                input_split_sizes=None,
+                group=self.device_group,
+            )
```
Comment (mingfeima) [inline python/sglang/multimodal_gen/runtime/distributed/device_communicators/xpu_communicator.py:186]: @zhangxiaoli73 also here, i don't quite get it why we need this Work around, not very decent.

File: python/sglang/multimodal_gen/runtime/distributed/device_communicators/xpu_communicator.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+            output = self._maybe_wait(output)
+            output = output.reshape(input_shape)
+
+            # Split and concatenate: [H, shard_seqlen, B, D] -> [shard_H, seqlen, B, D]
+            output = torch.cat(output.split(shard_hn), dim=1)
+
+            # Transpose back to [B, seqlen, shard_H, D]
+            output = output.transpose(0, 2).contiguous()
```
Comment (mingfeima) [inline python/sglang/multimodal_gen/runtime/distributed/device_communicators/xpu_communicator.py:194]: these will all end up with memcpy, be aware of the overhead.

File: python/sglang/multimodal_gen/runtime/distributed/device_communicators/xpu_communicator.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+            # Transpose to [shard_H, seqlen, B, D]
+            input_t = input_.transpose(0, 2).contiguous()
+
+            # Reshape for all-to-all: [shard_H * world_size, shard_seqlen, B, D]
+            input_t = (
+                input_t.reshape(shard_hn, self.world_size, shard_seqlen, bs, hd)
+                .transpose(0, 1)
+                .reshape(shard_hn * self.world_size, shard_seqlen, bs, hd)
+                .contiguous()
+            )
```
Comment (mingfeima) [inline python/sglang/multimodal_gen/runtime/distributed/device_communicators/xpu_communicator.py:213]:
```suggestion
input_t = torch.einsum('hwqbs->whqbs', input_t)  # swap first two dims
input_t = input_t.reshape(shard_hn * self.world_size, shard_seqlen, bs, hd)
```

reshape will trigger memcpy if not contiguous.

again, we should be cautious about these overheads.
anyway, you can leave it as it is right now, we can try to improve step by step later.

File: python/sglang/multimodal_gen/runtime/layers/attention/backends/intel_xpu.py [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+        causal,  # 19 bool     is_causal
+        window_size[0],  # 20 int      window_size_left
+        window_size[1],  # 21 int      window_size_right
+        softcap,  # 22 float    softcap
+        False,  # 23 bool     is_rotary_interleaved
+        None,  # 24 Tensor?  scheduler_metadata
+        num_splits,  # 25 int      num_splits
+        pack_gqa,  # 26 bool?    pack_gqa
+        sm_margin,  # 27 int      sm_margin
+    )
```
Comment (mingfeima) [inline python/sglang/multimodal_gen/runtime/layers/attention/backends/intel_xpu.py:165]: clean up the comments here.

File: python/sglang/multimodal_gen/runtime/layers/layernorm.py [hunk 7]
```diff
@@ -519,9 +552,16 @@ def tensor_parallel_rms_norm(x: torch.Tensor, norm: "RMSNorm") -> torch.Tensor:
...
+    # Use SUM + divide instead of AVG on XPU, because XCCL in torch 2.9+xpu have issues for AVG
+    if current_platform.is_xpu():
+        variance = get_tp_group().all_reduce(
+            variance, op=torch._C._distributed_c10d.ReduceOp.SUM
+        )
+        variance = variance / tp_size
+    else:
+        variance = get_tp_group().all_reduce(
+            variance, op=torch._C._distributed_c10d.ReduceOp.AVG
+        )
```
[the remark this answers is not in the document]
Comment (mingfeima) [reply python/sglang/multimodal_gen/runtime/layers/layernorm.py:564]: @xiangyuT

File: python/sglang/multimodal_gen/runtime/models/encoders/clip.py [hunk 8]
```diff
@@ -245,6 +245,57 @@ def forward(
                     is_causal=True,
                     scale=self.scale,
                 )
+            elif current_platform.is_xpu():
+                # XPU: Using both is_causal=True and attn_mask causes error:
+                # "attn_bias cannot present with is_causal"
```
Comment (mingfeima) [inline python/sglang/multimodal_gen/runtime/models/encoders/clip.py:250]: @liangan1

**Review by mingfeima — APPROVED** [review]
LGTM

**Comment by mingfeima** [comment]
generally LGTM, but this one does unveil several implementation limitations from torch distributed on XPU backend, we need to sync up internally to remove those work arounds gradually.

File: python/sglang/multimodal_gen/runtime/models/vaes/parallel/wan_dist_utils.py [hunk 9]
```diff
@@ -134,29 +134,48 @@ def halo_exchange(
...
-    if rank == world_size - 1:
-        recv_bottom_buf.zero_()
-
-    if p2p_ops:
-        reqs = dist.batch_isend_irecv(p2p_ops)
+    if current_platform.is_xpu():
+        # Use individual irecv/isend for XPU/XCCL compatibility
+        reqs = []
+
+        if rank > 0:
```
Comment (mingfeima) [inline python/sglang/multimodal_gen/runtime/models/vaes/parallel/wan_dist_utils.py:141]: no P2P batching

**Review by mingfeima — APPROVED** [review]

File: python/sglang/multimodal_gen/runtime/distributed/device_communicators/xpu_communicator.py [hunk 10]
```diff
@@ -0,0 +1,234 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (mingfeima) [inline python/sglang/multimodal_gen/runtime/distributed/device_communicators/xpu_communicator.py]:
```suggestion
```

we use this copy right claim, if you intend to add them:
```
/***************************************************************************************************
* Copyright (C) 2026 Intel Corporation, All rights reserved.
* Licensed under the MIT License (see LICENSE file)
**************************************************************************************************/
```

**Review by mingfeima — APPROVED** [review]

File: python/sglang/multimodal_gen/runtime/layers/attention/backends/intel_xpu.py (hunk 6 above)
Comment (xiangyuT) [reply python/sglang/multimodal_gen/runtime/layers/attention/backends/intel_xpu.py:165]: done

File: python/sglang/multimodal_gen/runtime/distributed/device_communicators/xpu_communicator.py (hunk 10 above)
Comment (xiangyuT) [reply python/sglang/multimodal_gen/runtime/distributed/device_communicators/xpu_communicator.py]: done

File: python/sglang/multimodal_gen/runtime/distributed/device_communicators/xpu_communicator.py (hunk 3 above)
Comment (zhangxiaoli73) [reply python/sglang/multimodal_gen/runtime/distributed/device_communicators/xpu_communicator.py:186]: In pytorch level, we have enabled `dist.all_to_all_single` unit tests and no issues are found. Could you please submit a ticket to public `torch-xpu-ops` if you meet problem?

File: python/sglang/multimodal_gen/runtime/models/vaes/parallel/wan_dist_utils.py (hunk 9 above)
Comment (zhangxiaoli73) [reply python/sglang/multimodal_gen/runtime/models/vaes/parallel/wan_dist_utils.py:141]: In pytorch level, we have enabled `batch_isend_irecv` unit tests and no issues are found. Could you please submit a ticket to public `torch-xpu-ops` if you meet problem?
